prog.go

```diff
@@ -629,3 +630,160 @@ func (p *BPFProg) DetachGenericFD(targetFd int, attachType BPFAttachType) error
+	Repeat      int
+	Duration    time.Duration
+	Flags       RunFlag
+	CPU         uint32
+	BatchSize   uint32
+}
+
+func runOptsToC(runOpts *RunOpts) (*C.struct_bpf_test_run_opts, error) {
+	if runOpts == nil {
+		return nil, nil
+	}
+
+	var (
+		dataIn      unsafe.Pointer
+		dataSizeIn  C.uint
+		dataOut     unsafe.Pointer
+		dataSizeOut C.uint
+		ctxIn       unsafe.Pointer
+		ctxSizeIn   C.uint
+		ctxOut      unsafe.Pointer
+		ctxSizeOut  C.uint
+	)
+
+	if runOpts.DataIn != nil {
+		dataIn = unsafe.Pointer(&runOpts.DataIn[0])
+		dataSizeIn = C.uint(runOpts.DataSizeIn)
+	}
+	if runOpts.DataOut != nil {
+		dataOut = unsafe.Pointer(&runOpts.DataOut[0])
+		dataSizeOut = C.uint(runOpts.DataSizeOut)
+	}
+	if runOpts.CtxIn != nil {
+		ctxIn = unsafe.Pointer(&runOpts.CtxIn[0])
+		ctxSizeIn = C.uint(runOpts.CtxSizeIn)
+	}
+	if runOpts.CtxOut != nil {
+		ctxOut = unsafe.Pointer(&runOpts.CtxOut[0])
+		ctxSizeOut = C.uint(runOpts.CtxSizeOut)
+	}
+	optsC, errno := C.cgo_bpf_test_run_opts_new(
+		dataIn,
+		dataOut,
+		dataSizeIn,
+		dataSizeOut,
+		ctxIn,
+		ctxOut,
+		ctxSizeIn,
+		ctxSizeOut,
+		C.int(runOpts.Repeat),
+		C.uint(runOpts.Flags),
+		C.uint(runOpts.CPU),
+		C.uint(runOpts.BatchSize),
+	)
+	if optsC == nil {
+		return nil, fmt.Errorf("failed to create bpf_test_run_opts: %w", errno)
+	}
+
+	return optsC, nil
+}
+
+func runOptsFromC(runOpts *RunOpts, optsC *C.struct_bpf_test_run_opts) {
+	if optsC == nil {
+		return
+	}
+
+	if optsC.data_in != nil {
+		runOpts.DataIn = C.GoBytes(optsC.data_in, C.int(optsC.data_size_in))
+	}
+	if optsC.data_out != nil {
+		runOpts.DataOut = C.GoBytes(optsC.data_out, C.int(optsC.data_size_out))
+	}
+	if optsC.ctx_in != nil {
+		runOpts.CtxIn = C.GoBytes(optsC.ctx_in, C.int(optsC.ctx_size_in))
+	}
+	if optsC.ctx_out != nil {
+		runOpts.CtxOut = C.GoBytes(optsC.ctx_out, C.int(optsC.ctx_size_out))
+	}
+
+	runOpts.RetVal = uint32(optsC.retval)
+	runOpts.Repeat = int(optsC.repeat)
+	runOpts.Duration = time.Duration(optsC.duration) * time.Nanosecond
+	runOpts.Flags = RunFlag(optsC.flags)
+	runOpts.CPU = uint32(optsC.cpu)
+	runOpts.BatchSize = uint32(optsC.batch_size)
+}
+
+// Run executes the eBPF program without attaching it to actual hooks, filling
+// the results in the provided RunOpts.
+// Reference:
+//   - https://docs.kernel.org/bpf/bpf_prog_run.html
+//   - https://docs.kernel.org/userspace-api/ebpf/syscall.html
+//
+// Example Usage:
+//
+//	/*
+//	SEC("tc")
+//	int test(struct __sk_buff *skb)
+//	{
+//	    return foo() ? 1 : 0;
+//	}
+//	*/
+//
+//	func TestFunc(t *testing.T) {
+//	    ...
+//	    prog, _ := module.GetProgram("test")
+//	    opts := RunOpts{
+//	        DataIn: make([]byte, 0, 14),
+//	        DataSizeIn: 14,
+//	        DataOut: make([]byte, 0, 14),
+//	        DataSizeOut: 14,
+//	        Repeat: 1,
+//	    }
+//	    prog.Run(&opts)
+//	    if opts.RetVal != 1 {
+//	        t.Errorf("result = %d; want 1", opts.RetVal)
+//	    }
+//	}
+func (p *BPFProg) Run(opts *RunOpts) error {
```

Comment:
I think it would be nice to add an error here when the program type isn't compatible (rather than relying on an internal failure).

Reply:
It will be dependent on kernel version support, e.g.:

- https://elixir.bootlin.com/linux/v6.9.3/source/net/bpf/test_run.c#L728 (supported)
- https://elixir.bootlin.com/linux/v6.9.3/source/include/linux/bpf.h#L3075 (unsupported)
- https://elixir.bootlin.com/linux/v6.9.3/source/include/linux/bpf.h#L2817 (unsupported)

The output for `ENOTSUPP` is `failed to run program: errno 524`. https://elixir.bootlin.com/linux/v6.9.3/source/include/linux/errno.h#L27

`.Error()` doesn't find a correct entry, since "not supported" for syscall is `ENOTSUP` 95. https://github.com/golang/sys/blob/master/unix/zerrors_linux_amd64.go#L614

Reply:
WDYT? If you agree, we must check all others returns for this `ENOTSUPP`.

```diff
diff --git a/prog.go b/prog.go
index e47c8ec..56ea57e 100644
--- a/prog.go
+++ b/prog.go
@@ -779,7 +779,11 @@ func (p *BPFProg) Run(opts *RunOpts) error {
 
        retC := C.bpf_prog_test_run_opts(C.int(p.FileDescriptor()), optsC)
        if retC < 0 {
-               return fmt.Errorf("failed to run program: %w", syscall.Errno(-retC))
+               errno := syscall.Errno(-retC)
+               if errno == 524 { // ENOTSUPP https://elixir.bootlin.com/linux/v6.9.3/source/include/linux/errno.h#L27
+                       errno = syscall.ENOTSUP
+               }
+               return fmt.Errorf("failed to run program: %s", errno)
        }
 
        // update runOpts with the values from the kernel and libbpf
```

Reply:
Anyway, merging. If decided for it, we can tackle it in a bunch.